File: backend/app/schemas/resume.py
import uuid
from datetime import date, datetime

from pydantic import BaseModel, Field


# --- Basic Info ---

class BasicInfo(BaseModel):
    name: str = ""
    phone: str = ""
    email: str = ""
    address: str = ""
    website: str = ""
    linkedin: str = ""


# --- Education ---

class EducationBase(BaseModel):
    school: str = Field(min_length=1, max_length=200)
    degree: str | None = None
    major: str = Field(min_length=1, max_length=200)
    start_date: date | None = None
    end_date: date | None = None
    gpa: str | None = None
    description: str | None = None


class EducationCreate(EducationBase):
    pass


class EducationUpdate(BaseModel):
    school: str | None = None
    degree: str | None = None
    major: str | None = None
    start_date: date | None = None
    end_date: date | None = None
    gpa: str | None = None
    description: str | None = None


class EducationResponse(EducationBase):
    id: uuid.UUID
    resume_id: uuid.UUID

    model_config = {"from_attributes": True}


# --- Work Experience ---

class WorkExperienceBase(BaseModel):
    company: str = Field(min_length=1, max_length=200)
    title: str = Field(min_length=1, max_length=200)
    location: str | None = None
    start_date: date | None = None
    end_date: date | None = None
    current: bool = False
    description: str | None = None


class WorkExperienceCreate(WorkExperienceBase):
    pass


class WorkExperienceUpdate(BaseModel):
    company: str | None = None
    title: str | None = None
    location: str | None = None
    start_date: date | None = None
    end_date: date | None = None
    current: bool | None = None
    description: str | None = None


class WorkExperienceResponse(WorkExperienceBase):
    id: uuid.UUID
    resume_id: uuid.UUID

    model_config = {"from_attributes": True}


# --- Project ---

class ProjectBase(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    role: str | None = None
    start_date: date | None = None
    end_date: date | None = None
    description: str | None = None
    highlights: list[str] | None = None
    tech_stack: list[str] | None = None
    project_url: str | None = None


class ProjectCreate(ProjectBase):
    pass


class ProjectUpdate(BaseModel):
    name: str | None = None
    role: str | None = None
    start_date: date | None = None
    end_date: date | None = None
    description: str | None = None
    highlights: list[str] | None = None
    tech_stack: list[str] | None = None
    project_url: str | None = None


class ProjectResponse(ProjectBase):
    id: uuid.UUID
    resume_id: uuid.UUID

    model_config = {"from_attributes": True}


# --- Skill ---

class SkillBase(BaseModel):
    category: str | None = None
    name: str = Field(min_length=1, max_length=100)
    level: str | None = None


class SkillCreate(SkillBase):
    pass


class SkillUpdate(BaseModel):
    category: str | None = None
    name: str | None = None
    level: str | None = None


class SkillResponse(SkillBase):
    id: uuid.UUID
    resume_id: uuid.UUID

    model_config = {"from_attributes": True}


# --- Resume ---

class ResumeCreate(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    target_position: str | None = None
    language: str = "zh"


class ResumeUpdate(BaseModel):
    title: str | None = None
    basic_info: BasicInfo | None = None
    summary: str | None = None
    target_position: str | None = None
    language: str | None = None


class ResumeResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    title: str
    basic_info: dict | None = None
    summary: str | None = None
    target_position: str | None = None
    language: str
    guided_data: dict | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ResumeDetailResponse(ResumeResponse):
    education: list[EducationResponse] = []
    work_experiences: list[WorkExperienceResponse] = []
    projects: list[ProjectResponse] = []
    skills: list[SkillResponse] = []

File: frontend/src/types/resume.ts
export interface BasicInfo {
  name: string;
  phone: string;
  email: string;
  address: string;
  website: string;
  linkedin: string;
}

export interface ResumeEducation {
  id: string;
  resume_id: string;
  school: string;
  degree: string | null;
  major: string;
  start_date: string | null;
  end_date: string | null;
  gpa: string | null;
  description: string | null;
}

export interface ResumeWorkExperience {
  id: string;
  resume_id: string;
  company: string;
  title: string;
  location: string | null;
  start_date: string | null;
  end_date: string | null;
  current: boolean;
  description: string | null;
}

export interface ResumeProject {
  id: string;
  resume_id: string;
  name: string;
  role: string | null;
  start_date: string | null;
  end_date: string | null;
  description: string | null;
  highlights: string[] | null;
  tech_stack: string[] | null;
  project_url: string | null;
}

export interface ResumeSkill {
  id: string;
  resume_id: string;
  category: string | null;
  name: string;
  level: string | null;
}

export interface Resume {
  id: string;
  user_id: string;
  title: string;
  basic_info: BasicInfo | null;
  summary: string | null;
  target_position: string | null;
  language: string;
  created_at: string;
  updated_at: string;
}

export interface ResumeDetail extends Resume {
  education: ResumeEducation[];
  work_experiences: ResumeWorkExperience[];
  projects: ResumeProject[];
  skills: ResumeSkill[];
}

export interface ChatMessage {
  id: string;
  role: "user" | "ai";
  content: string;
  timestamp: number;
}

export interface EducationInput {
  school: string;
  degree?: string;
  major: string;
  start_date?: string;
  end_date?: string;
  gpa?: string;
  description?: string;
}

export interface WorkExperienceInput {
  company: string;
  title: string;
  location?: string;
  start_date?: string;
  end_date?: string;
  current?: boolean;
  description?: string;
}

export interface ProjectInput {
  name: string;
  role?: string;
  start_date?: string;
  end_date?: string;
  description?: string;
  highlights?: string[];
  tech_stack?: string[];
  project_url?: string;
}

export interface SkillInput {
  category?: string;
  name: string;
  level?: string;
}
